cli/azd/cmd/container.go

```diff
@@ -179,13 +179,13 @@ func registerCommonDependencies(container *ioc.NestedContainer) {
 	})
 
 	// Azd Context
-	container.MustRegisterSingleton(azdcontext.NewAzdContext)
+	container.MustRegisterSingleton(func(lazyAzdContext *lazy.Lazy[*azdcontext.AzdContext]) (*azdcontext.AzdContext, error) {
```

Comment:
@ellismg I believe these changes get us closer to what you were seeing where we will be better off resolving the instance from the shared lazy instead of the other way around.

Reply:
Is it OK to think that, whenerver a `azdcontext.AzdContext` is requested from the container, the container will first create a lazy[AzdContext] to gets the azdContext from?  

Can you help me understand:
- When do I want to inject an AzdContext v/s a Lazy[AzdContext].  I feel like, with this change, everything is updated to use Lazy[AzdContext] ?  (becase AzdContext is becoming a pointer to a lazy[AzdContext]) ?

Reply:
The use case is to still use the standard `AzdContext` to be injected in most situations.  The only times we should be using the lazy is for the conditions where the AzdContext may not yet be available and need to be re-evaluated later in the component.

The idea here is that our IoC container will take a nil return and cache it for future resolutions.  This isn't ideal and for the components like AzdContext, ProjectConfig and other similar constructs if we depend on the Lazy[T] then it will allow the resolver to always return the same lazy value that may or not have a value.  Then when the lazy does return a valid value other callers will have the correct instance.  I will create a couple unit tests to illustrate this point.